src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums, List<int> list)
+    {
+        list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums, List<int> list)
+    {
+        list = (from int n1 in nums
+                from int n2 in nums
+                select n1).ToList();
+        return list;
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListToArrayElement()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                lists[0].Add(n1);
+            }
+        }|]
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0].AddRange(from int n1 in nums
+                          from int n2 in nums
+                          select n1);
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListToNewArrayElement()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0] = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                lists[0].Add(n1);
+            }
+        }|]
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0] = (from int n1 in nums
+                    from int n2 in nums
+                    select n1).ToList();
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListHashSetNoConversion()
+        {
+            string source = @"
+using System.Collections.Generic;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        var hashSet = new HashSet<int>();
+        [|foreach (int n1 in nums)
+        {
+            hashSet.Add(n1);
+        }|]
```

Comment:
your design seems flexible enough to support this kind of thing really easy if you want to. very nice!
hashSet.UnionWith( ... ) rather than list.AddRange( ... )

Reply:
Thank you! Added to [https://github.com/dotnet/roslyn/issues/27980](https://github.com/dotnet/roslyn/issues/27980)

---
In reply to: [195856493](https://github.com/dotnet/roslyn/pull/27130#discussion_r195856493) [](ancestors = 195856493)